### DISCUSSION TITLE
MacOS and Linux Testing

### DISCUSSION THREAD

**Discussion by Deijin27** [Deijin27/RanseiLink#51] [General]
A discussion for testing out AvaloniaUI's viablility for supporting a cross-platform version of ranseilink

If you have a Mac computer and are willing to help out, please post a message here. (This requires no expertise, it will just be me giving you an app to run and reply back with whether it works and screenshots of how it looks).

I'm currently capable of testing on Linux x64 via:
- Ubuntu (Windows Subsystem for Linux)
- Mint (Running via a bootable usb, and I have an old windows laptop that I'm considering installing linux on)

But of course, anyone with linux, especially different distros or arm64 etc., are welcome to help test.

Prerelease builds here:
https://github.com/Deijin27/RanseiLink/releases/tag/v6.0-pre

<details>
<summary>Verifications</summary>

| Test | Windows | Mac | Linux |
|-------|---------|-----|-------|
| RanseiLink logo is on unzipped executable / appbundle | yes | ? | no [3] |
| Launches [1] | yes | ? | yes |
| Controls look right and work correctly | yes | ? | yes |
| Right-click menu on textbox works | yes | ? | yes |
| RanseiLink logo is on taskbar | yes | ? | yes |
| Can open and close import mod dialog | yes | ? | yes |
| Can use file picker to select a mod to import | yes | ? | yes |
| Can use drag-drop to select a mod to import | yes | ? | no [2] |
| Mod is imported successfully | yes | ? | yes |
| Right click mod to open in file explorer works | yes | ? | yes |
| Where are the mods located? | `%LOCALAPPDATA%/RanseiLink` | ? | `/home/mint/.local/share/RanseiLink` |
| Can use folder picker to choose where to export | yes | ? | yes |
| Create mod works | yes | ? | yes |
| Edit mod info works | yes | ? | yes |
| Create mod based on works | yes | ? | yes |
| Delete mod works | yes | ? | yes |
| Populate graphics defaults works (verify in `.../RanseiLink/DefaultData/VPYT/graphics`) | yes | ? | yes |
| Patch mod works (make sure to use one with modified images and test by running the patched rom) | yes | ? | yes |
| Toggle theme works | yes | ? | yes |
| Remembers window dimensions on restart | yes | ? | yes |
| Bug reporter launches and the "version" field is filled in on the webpage | yes | ? | yes |
| Crash reporter launches and the "version" and "error message" fields are filled in on the webpage | yes | ? | yes |

[1] linux and mac should not require chmod, this permission is preserved when the file is zipped  
[2] known issue in the avalonia repo that drag-drop files doesn't work on linux
[3] idk how this works on linux, is it possible?

</details>

**Comment by Deijin27** [comment]
With a prototype I've been testing the linux x64 build

<details>
<summary>Lots of stuff</summary>

## Windows Version for comparison: on Dell XPS 15 Windows 11

![Screenshot (61)](https://github.com/Deijin27/RanseiLink/assets/40903783/9cc2c25b-e575-426f-bc7f-a964eddc3818)

## WSL Ubuntu, on Dell XPS 15 Windows 11

![Screenshot (60)](https://github.com/Deijin27/RanseiLink/assets/40903783/0a104432-419f-4f7a-9eeb-dffb5b422c86)

- It runs fine directly in wsl assuming you've run `sudo chmod +x RanseLink.XP`
- I think the scaling is messed up, because everything looks small and blurry compared to windows version.
- Some of the text is black, I think it's because Avalonia uses default system text color. When setting text color explicitly in all cases, it works correctly.
- The dialog box left side looks paler because I accidentally didn't remove transparency from the background color
- Mouse-down on the combobox makes it have a black border, most likely similar issues to text, but I need to investigate to be sure.
- Updown buttons on number box should have a line between them and beside, constructed by two 0.5 pixel lines, This implies that using half pixels is not an option and the styles need to be adapted, maybe making the compromise of a double width between buttons, or a missing side on one of the buttons.
- Sometimes ui freezes until you move mouse
- No context menu on text box

## Linux Mint, Bootable USB on Dell Vostro 15

![Screenshot from 2023-08-05 16-29-50](https://github.com/Deijin27/RanseiLink/assets/40903783/cd6a509c-f3b2-4808-b2aa-15262f5e7b94)

- Ignore the extra black text, this is an older build than I was running on WSL
- The scaling is much better, confirming it's a WSL issue, same with ui freezing issues, and the context menus work on text boxes too
- The issue that remains is the missing 0.5 pixel lines, so this is something that will need addressing.

</details>

## What's next

Next up I want to test:
- where the mods are created, and that the load correctly. On WSL they're created in the folder with the app, this is fine tbh, but I haven't tested on mint yet, it might be that the "special folders" work on a real OS.
- I just remembered there was an issue with scroll bars on WSL, need to test that on mint
- And test if openTK works, so we could use the 3d model viewer I wrote

**Reply by Deijin27** [reply]
As an update to this, i tested and the mods are created in the "special folder" that .net uses for localappdata on linux, so tgats good.

I also tested all file interactions and its all working good. So it's in a functional stated to be used to patch a rom with mods made by other people.

Still need to try openTK.

I've found a nice way to make the wpf dialogservice async, so i should be able to swap out all them and them pretty much all view models can be moved to the GuiCore project and shared so incremental development of xp can start without need for duplicating view models.

**Comment by Baconfry** [comment]
I just factory reset my old Mac to restore it to a usable state, but as a result, it is using a very old version of MacOS (10.11.6 El Capitan). I doubt I will be updating it further because unlike Apple, I know better than to try to fix things when they aren't broken. I do not know how applicable my results will be for modern versions, but I can try to help anyways

**Reply by Deijin27** [reply]
Thanks for offering to help. That version should be high enough. I'm particularly interested in the behaviour of Gatekeeper, which was introduced in 10.7.3.

Things that are priorities to test:

- Does it run? I know simlar to linux it needs `chmod +x` applying which can be done on my end. Has the app bundle I create got everything it needs, and is it blocked by gatekeeper, how does that blocking actually manifest?
- Does the interface scale fine, look right, can be interacted with with no issues?
- Where does it store "special folders" for localappdata (where mods, settings, default data is stored) and temp (where data is put during patching)? Can it read and write to these locations ok, and also for files elsewhere i.e. the rom?

I'll get together an app bundle and add a few more basic features to the tool, then send it your way.

**Reply by Deijin27** [reply]
Technically the notarisation requirement was added in 10.15, but i wont be signing either because it requires a mac computer to run the codesign tool that comes with xcode. And definitely not notarising as that requires a paid developer account.

So I'm thinking people are gonna have to workaround gatekeeper with 

https://support.apple.com/en-us/HT202491

Section called "If you want to open an app that hasn’t been notarized or is from an unidentified developer".

**Reply by Deijin27** [reply]
So I've got a [build together](https://github.com/Deijin27/RanseiLink/releases/tag/v6.0-pre), and all the functionality is correctly working on linux. I've tried my best with getting the app bundle set up correctly, but you can never know until you try it of course.

I did notice an issue that the [avalonia requirements](https://github.com/AvaloniaUI/Avalonia/wiki/Runtime-Requirements) says it needs high sierra + to work. I think it's still worth trying on the older version that you have, and if it doesn't work we can look for someone else to test.

**Comment by DarthMDev** [comment]
i have an m3 mac i can help you test on

**Reply by DarthMDev** [reply]
so i gave the pre release build a try on arm and moved it to applications , 
"The application “RanseiLink.app” can’t be opened."

**Reply by Deijin27** [reply]
Thanks! They haven't really given us much to work with from that error message :(

When you put it in the apps folder, the location where you launch it, what does it look like? E.g. is there an icon, a name and info about it?

Theres also a possibility it doesnt have execution permissions, I did give it them but they may not have persisted, not sure if thats the error you would get from it though.

I'll have another check over it to see if I can spot issues. Or maybe figure out a way to test it in a more bare bones way incase I've made a mistake while constructing the app package around it.

**Reply by DarthMDev** [reply]
Yeah i tried chmod +x on all the executables in there nothing, tried running the unix executable in there directly, nothing. Very strange. This is also after doing the open anyway because apple cant verify it for malware or whatever

**Reply by Deijin27** [reply]
Could you try one of these apps, made with the same ui framework. If  they launch we know its possible to work on your pc and I'm doing something wrong / they want me to pay the fruit dollars to sign and notarise.

This one has some interesting info in the readme specific to mac:
https://github.com/sourcegit-scm/sourcegit

This one has a signed and notarised version:
https://github.com/carina-studio/ULogViewer

More example apps here:
https://avaloniaui.net/Showcase

**Comment by OrdinaryCloverOC** [comment]
i can test on fedora

**Reply by OrdinaryCloverOC** [reply]
(also yall got 6.1?)
